discretize/tests.py

```diff
@@ -633,6 +637,110 @@ def Quadratic(x, return_g=True, return_H=True):
     return Quadratic
 
 
+# TESTS
+def dottest(
+    forward,
+    adjoint,
+    fwd_shape,
+    adj_shape,
+    fwd_complex=False,
+    adj_complex=False,
+    fwd_order="C",
+    adj_order="C",
+    clinear=True,
+    rtol=1e-6,
+    atol=0.0,
+    raise_error=False,
+    verb=False,
+):
+    r"""Dot product test for the forward operator and its adjoint operator.
+
+    Dot product test to verify the correctness of the adjoint operator
+    :math:`F^H` of the forward operator :math:`F`.
+
+    .. math::
+
+        \mathbf{v}^H ( \mathbf{F} \mathbf{u} ) =
+        ( \mathbf{F}^H \mathbf{v} )^H \mathbf{u}
+
+
+    Parameters
+    ----------
+    forward, adjoint : functions
+        Forward operator and its adjoint operator.
+
+    fwd_shape, adj_shape : int, tuple of int
+        Shapes of the vectors {u; v} passed to ``forward`` and ``adjoint``.
+
+    fwd_complex, adj_complex : bool, defaults: False, False
+        If True, complex vectors are passed to ``forward`` and ``adjoint``.
+
+    fwd_order, adj_order : string, defaults: 'C', 'C'
+        Used the ravel the output of the ``forward``/``adjoint`` if their shape
+        is higher than 1D.
+
+    clinear : bool, default: True
+        If operator is complex-linear (True) or real-linear (False).
+
+    rtol, atol : float, defaults: 1e-6, 0.0
+        Relative and absolute tolerance, used with :func:`numpy.isclose`.
+
+    raise_error : bool, default: False
+        If False, the result of the test is returned as boolean.
+        If True and the test fails, an AssertionError is raised.
+
+    verb : bool, default: False
+        Verbosity; if True, print result if passed.
+
+
+    Returns
+    -------
+    passed : bool, returned if raise_error=False
+        Result of the dot product test.
+
+    """
+
+    def random(size, iscomplex):
+        """Create random data of size and dtype of <size>."""
+        out = rng.standard_normal(int(np.real(size)))
+        if iscomplex:
+            out = out + 1j * rng.standard_normal(out.size)
+        return out
+
+    # Create random vectors u and v.
+    u = random(np.product(fwd_shape), fwd_complex)
+    v = random(np.product(adj_shape), adj_complex)
+
+    # Carry out dot product test.
+    fwd_u = forward(u.reshape(fwd_shape, order=fwd_order)).ravel(fwd_order)
```

Comment:
This won't work if the `forward` and the `adjoint` take different shapes.

Reply:
As long as the out put of the forward matches adj_shape, and the output of adj match’s forward shape. It will work. The two functions do not need to accept the same shaped argument.

Reply:
This is/should be true indeed. I had a test example where it failed, but probably my example was bonkers! I'll look at it again.

Reply:
If the functions returned the appropriate shaped arrays, you could even flatten them out irrespective of their internal ordering, as long as they're both flattened out with the same ordering.